controllers/drcluster_controller.go

```diff
@@ -266,17 +280,32 @@ func (u *drclusterUpdater) statusConditionSet(
 	return nil
 }
 
-func (u *drclusterUpdater) statusUpdate() error {
-	return u.client.Status().Update(u.ctx, u.object)
+func (u *drclusterInstance) statusUpdate() error {
+	if !reflect.DeepEqual(u.savedInstanceStatus, u.object.Status) {
+		if err := u.client.Status().Update(u.ctx, u.object); err != nil {
+			u.log.Info(fmt.Sprintf("Failed to update drCluster status (%s/%s/%v)",
+				u.object.Name, u.object.Namespace, err))
+
+			return fmt.Errorf("failed to update drCluster status (%s/%s)", u.object.Name, u.object.Namespace)
+		}
+
+		u.log.Info(fmt.Sprintf("Updated drCluster Status %+v", u.object.Status))
+
+		return nil
```

Comment:
statusUpdate is called from multiple places, if the first caller determined that the status should be updated (IOW DeepEqual fails), then a subsequent caller who does not change anything in the status would also attempt to inadvertently update the status as the DeepEqual of current instance status and saved status would mismatch due to the first caller.

We need to stash the updated status into saved status, or update status once in the end.

The latter above would be preferred, as once status is updated the object revision changes and a subsequent status update to the object would result in an error as the server resource is at a later version.

Reply:
statusUpdate is called from 2 places as of now.

1) From the function statusConditionSet which itself is called from validatedSetFalse.

2) processFencing

(1) happens when validation of drCluster fails. If that happens, then processFencing is not executed at all and the reconciler requeues.

If (1) succeeds, then validatedSetFalse function is not called. Just the structure's condition of DRClusterValidated is set to true and (2) is called. And it is (2) where statusUpdate gets executed. 

Hence even though there are 2 places where statusUpdate is invoked, at any particular reconcilation only one of them will be invoked, not both.

Let me know if I am missing something or interpreted the comment differently.

Reply:
Someone has to remember this in the future in case of changes. It is better to update status once in the end, and cache all intermediate changes.

Reply:
Ok so the function validatedSetTrue is removed in an older PR, so this drifts from DRPolicy based status updates, which causes the confusion.

At present this is fine, as we only update the validated status in memory and update the API resource towards the end.